metagpt/tools/code_interpreter.py

```diff
@@ -0,0 +1,128 @@
+import re
+from typing import List, Callable
+from pathlib import Path
+
+import wrapt
+import textwrap
+import inspect
+from interpreter.interpreter import Interpreter
+
+from metagpt.logs import logger
+from metagpt.config import CONFIG
+from metagpt.utils.highlight import highlight
+from metagpt.actions.clone_function import CloneFunction, run_fucntion_code, run_fucntion_script
+from metagpt.actions.run_code import RunCode
+
+
+def extract_python_code(code: str):
+    """提取代码块: 如果代码注释相同，则只保留最后一个代码块."""
+    # 使用正则表达式匹配注释块和相关的代码
+    pattern = r'(#\s[^\n]*)\n(.*?)(?=\n\s*#|$)'
+    matches = re.findall(pattern, code, re.DOTALL)
+
+    # 提取每个相同注释的最后一个代码块
+    unique_comments = {}
+    for comment, code_block in matches:
+        unique_comments[comment] = code_block
+
+    # 组装结果字符串
+    result_code = '\n'.join([f"{comment}\n{code_block}" for comment, code_block in unique_comments.items()])
+    header_code = code[:code.find("#")]
+    code = header_code + result_code
+
+    logger.info(f"Extract python code: \n {highlight(code)}")
+
+    return code
+
+
+class OpenCodeInterpreter(object):
+    """https://github.com/KillianLucas/open-interpreter"""
+    def __init__(self, auto_run: bool = True) -> None:
+        interpreter = Interpreter()
+        interpreter.auto_run = auto_run
+        interpreter.model = CONFIG.openai_api_model or "gpt-3.5-turbo"
+        interpreter.api_key = CONFIG.openai_api_key
```

Comment:
Would it be better to initialize the interpreter instance directly using these parameters?

Reply:
As of the current version 0.1.3, [Interpreter](https://github.com/KillianLucas/open-interpreter/blob/main/interpreter/interpreter.py)  only supports initialize the interpreter instance directly using these parameters. Here is the code for Interpreter
```python

class Interpreter:

  def __init__(self):
    self.messages = []
    self.temperature = 0.001
    self.api_key = None
    self.auto_run = False
    self.local = False
    self.model = "gpt-4"
    self.debug_mode = False
    ...
```